libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c

```diff
@@ -376,166 +278,591 @@ CK_RV prvMbedTLS_Initialize( void )
-                                            CK_SLOT_ID_PTR pxSlotList,
-                                            CK_ULONG_PTR pulCount )
-{   /*lint !e9072 It's OK to have different parameter name. */
+CK_RV prvDeleteObjectFromList( CK_OBJECT_HANDLE xAppHandle )
+{
     CK_RV xResult = CKR_OK;
+    BaseType_t xGotSemaphore;
+    int lIndex = xAppHandle - 1;
 
-    /* Since the mbedTLS implementation of PKCS#11 does not depend
-     * on a physical token, this parameter is ignored. */
-    ( void ) ( xTokenPresent );
+    xGotSemaphore = xSemaphoreTake( xP11Context.xObjectList.xMutex, portMAX_DELAY );
 
-    if( NULL == pulCount )
-    {
-        xResult = CKR_ARGUMENTS_BAD;
-    }
-    else if( NULL == pxSlotList )
-    {
-        *pulCount = 1;
-    }
-    else
+    if( xGotSemaphore == pdTRUE )
     {
-        if( 0u == *pulCount )
+        if( xP11Context.xObjectList.xObjects[ lIndex ].xHandle != CK_INVALID_HANDLE )
         {
-            xResult = CKR_BUFFER_TOO_SMALL;
+            memset( &xP11Context.xObjectList.xObjects[ lIndex ], 0, sizeof( P11Object_t ) );
         }
         else
         {
-            pxSlotList[ 0 ] = pkcs11SLOT_ID;
-            *pulCount = 1;
+            xResult = CKR_OBJECT_HANDLE_INVALID;
         }
+
+        xSemaphoreGive( xP11Context.xObjectList.xMutex );
+    }
+    else
+    {
+        xResult = CKR_CANT_LOCK;
     }
 
     return xResult;
 }
 
+
 /**
- * @brief Start a session for a cryptographic command sequence.
- */
-CK_DEFINE_FUNCTION( CK_RV, C_OpenSession )( CK_SLOT_ID xSlotID,
-                                            CK_FLAGS xFlags,
-                                            CK_VOID_PTR pvApplication,
-                                            CK_NOTIFY xNotify,
-                                            CK_SESSION_HANDLE_PTR pxSession )
-{   /*lint !e9072 It's OK to have different parameter name. */
+* @brief Add an object that exists in NVM to the application object array.
+*
+* @param[in[ xPalHandle         The handle used by the PKCS #11 PAL for object.
+* @param[out] pxAppHandle       Updated to contain the application handle corresponding to xPalHandle.
+* @param[in]  pcLabel           Pointer to object label.
+* @param[in] xLabelLength       Length of the PKCS #11 label.
+*
+*/
+CK_RV prvAddObjectToList( CK_OBJECT_HANDLE xPalHandle,
+                          CK_OBJECT_HANDLE_PTR pxAppHandle,
+                          uint8_t * pcLabel,
+                          size_t xLabelLength )
+{
     CK_RV xResult = CKR_OK;
-    P11SessionPtr_t pxSessionObj = NULL;
+    BaseType_t xGotSemaphore;
 
-    ( void ) ( xSlotID );
-    ( void ) ( pvApplication );
-    ( void ) ( xNotify );
+    CK_BBOOL xObjectFound = CK_FALSE;
+    int lInsertIndex = -1;
+    int lSearchIndex = pkcs11configMAX_NUM_OBJECTS - 1;
 
-    /* Check arguments. */
-    if( NULL == pxSession )
+    xGotSemaphore = xSemaphoreTake( xP11Context.xObjectList.xMutex, portMAX_DELAY );
```

Comment:
"destroy an object in the middle of an operation" - that language seems to suggest intention.  Is there anything to prevent an unintentional case where one task is using an object that another task is destroying?  If not, we might want to provide warnings around the use of destroy.